sdks/python/apache_beam/transforms/window.py

```diff
@@ -246,10 +263,33 @@ def __init__(self, value, timestamp):
     self.value = value
     self.timestamp = Timestamp.of(timestamp)
 
-  def __cmp__(self, other):
-    if type(self) is not type(other):
-      return cmp(type(self), type(other))
-    return cmp((self.value, self.timestamp), (other.value, other.timestamp))
+  def __eq__(self, other):
```

Comment:
Using `total_ordering` results in unexpected behavior. Concretely the test `test_reshuffle_windows_unchanged` fails. 

I have tried to locate the exact cause by implementing all OPs (with the `total_ordering` decorator in place) and subsequently leaving out the OPs one by one:

1) adding the `total_ordering` decorator itself doesn't introduce issues
2) only use `total_ordering` to fill in  `__lt__` works, using other combinations always fail.
3) I am currently testing the OPs by manually using the conversion rules defined by [total_ordering](https://github.com/python/cpython/blob/master/Lib/functools.py) to see if I can locate the exact problem.

@tvalentyn: Given [the note](https://docs.python.org/3/library/functools.html#functools.total_ordering) on performance impact of the total_ordering decorator, it might make sense to implement all OPs instead of using the decorator? That works already, in the meantime I will continue the testing (step3) to see if I can give more info.

Reply:
With this PR, the test becomes flaky, or in other words passes sometimes. It may still flake if we implement all ops manually - did you try running it multiple times?

I don't understand yet what change in behavior triggers this (we should find out), but I think we need to fix the test regardless of this: https://github.com/apache/beam/pull/6104.

Performance-wise, last week I used https://github.com/apache/beam/compare/master...tvalentyn:transforms_microbenchmark to compare different options, and did not notice a significant difference when using @total_ordering (we could double check), so I favored the decorator to reduce the boilerplate.

Reply:
Yes, I just retested with the full implementation which seems to work again. However, will be good to test the @total_ordering after your PR has been merged :-).

Reply:
I checked performance of `windowed_value`, `interval_window`, `timestamped_value`, `bounded_window` in dictionaries and ordered lists, with and without this PR. For the most part, performance is not changed or  improved. `@total_ordering` does not significantly affect it. Only concern is using `hash(type(self))` when evaluating hashes of objects may be unnecessary in most cases, and slightly decreases the performance here: https://github.com/apache/beam/pull/5729/files#diff-d7dfd884622fb59806ba9276cf3bd8fbR242. So I left some more comments to simplify hash functions. The change above was also the trigger for test flakiness, although ultimately the test was at fault. 

Without PR:
```
wv_with_one_window: dict, 10000 element(s)      : per element median time cost: 4.71699e-06 sec, relative std: 5.93%						
wv_with_multiple_windows: dict, 10000 element(s): per element median time cost: 4.02698e-05 sec, relative std: 0.60%						
interval_window: dict, 10000 element(s)         : per element median time cost: 1.5276e-06 sec, relative std: 1.78%						
timestamped_value: dict, 10000 element(s)       : per element median time cost: 1.39499e-07 sec, relative std: 7.44%						
interval_window: sorting., 10000 element(s)     : per element median time cost: 4.04392e-05 sec, relative std: 0.63%						
timestamped_value: sorting., 10000 element(s)   : per element median time cost: 1.80363e-05 sec, relative std: 1.35%						
bounded_window: sorting., 10000 element(s)      : per element median time cost: 4.06633e-05 sec, relative std: 1.26%						
```
With PR (including the change suggested in last iteration).

```
wv_with_one_window: dict, 10000 element(s)      : per element median time cost: 5.047e-06 sec, relative std: 2.16%
wv_with_multiple_windows: dict, 10000 element(s): per element median time cost: 4.0575e-05 sec, relative std: 2.20%
interval_window: dict, 10000 element(s)         : per element median time cost: 1.53821e-06 sec, relative std: 2.43%
timestamped_value: dict, 10000 element(s)       : per element median time cost: 1.27995e-06 sec, relative std: 6.11%
interval_window: sorting., 10000 element(s)     : per element median time cost: 1.83087e-05 sec, relative std: 1.28%
timestamped_value: sorting., 10000 element(s)   : per element median time cost: 8.4375e-06 sec, relative std: 2.62%
bounded_window: sorting., 10000 element(s)      : per element median time cost: 1.80462e-05 sec, relative std: 3.56%
```